### PR TITLE
Ip restrictions

### PR DESCRIPTION
**Pull request by asif-khan17**
**Fixed Issues**

#23040,#23041,#23042,#23043,#23044,#23046

**Files Changed**

TestController.cs
TestAC.cs - Modified
TestIPAddressAC.cs - Added
TrappistDbContext.cs
Test.cs
ITestsRepository.cs
TestsRepository.cs
TestConductRepository.cs
TestConductRepositoryTest.cs
TestRepositoryTest.cs
tests.service.ts
tests.settings.html
tests.settings.component.ts
testIPAddress.ts-  Added
questions-dashboard.component.ts
questions-dashboard.html
questions-single-multiple-answer.component.ts
questions-single-multiple-answer.component.html
single-multiple-answer-question-option.model.ts
create-test-footer.component.ts
create-test-footer.html
create-test-header.component.ts
create-test-header.html
Bootstrap.cs

**Checks**

- [X] Naming Conventions
- [X] Server side code comments (proper English, grammar and no spelling mistakes)
- [X] Client side code comments (proper English, grammar and no spelling mistakes)
- [X] Unused variables, methods, blank spaces and name spaces. 
- [X] Optimal Code and code formatting
- [X] Commit History is proper, no merge is done. 
- [X] Commit/pull request messages should be proper to justify your feature
- [ ] All test cases are executed provided by tester.

**Known Issues**

- issue 1
- issue 2

**Comments**

NA

---
This change is [Reviewable](https://reviewable.io/reviews/promact/trappist/387)

### CODE REVIEW

File: Trappist/src/Promact.Trappist.DomainModel/ApplicationClasses/Test/TestIPAddressAC.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+﻿
+using System;
+
+namespace Promact.Trappist.DomainModel.ApplicationClasses.Test
+{
+    public class TestIPAddressAC
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.DomainModel/ApplicationClasses/Test/TestIPAddressAC.cs]: MINOR Rename class 'TestIPAddressAC' to match camel case naming rules, consider using 'TestIpAddressAC'. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS101)

File: Trappist/src/Promact.Trappist.DomainModel/ApplicationClasses/Test/TestIPAddressAC.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        public int Id { get; set; }
+
+        public string IPAddress { get; set; }
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.DomainModel/ApplicationClasses/Test/TestIPAddressAC.cs]: MINOR Rename property 'IPAddress' to match camel case naming rules, consider using 'IpAddress'. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS100)

File: Trappist/src/Promact.Trappist.Core/Controllers/TestController.cs [hunk 3]
```diff
@@ -120,11 +120,18 @@ public TestsController(ITestsRepository testRepository, IStringConstants stringC
...
-        public async Task<IActionResult> PauseTest([FromRoute] int id,[FromRoute] bool isPause)
+        public async Task<IActionResult> PauseTestAsync([FromRoute] int id,[FromRoute] bool isPause)
         {
-            await _testRepository.PauseResumeTest(id,isPause);
+            await _testRepository.PauseResumeTestAsync(id,isPause);
             return Ok(isPause);
         }
+
+        [HttpDelete("deleteIp/{id}")]
+        public async Task<IActionResult> DeleteIPAsync([FromRoute] int id)
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.Core/Controllers/TestController.cs]: MINOR Rename method 'DeleteIPAsync' to match camel case naming rules, consider using 'DeleteIpAsync'. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS100)

File: Trappist/src/Promact.Trappist.DomainModel/ApplicationClasses/Test/TestAC.cs [hunk 4]
```diff
@@ -47,6 +48,8 @@ public class TestAC
 
         public List<CategoryAC> CategoryAcList { get; set; }
 
+        public List<TestIPAddressAC> TestIPAddress { get; set; }
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.DomainModel/ApplicationClasses/Test/TestAC.cs]: MINOR Rename property 'TestIPAddress' to match camel case naming rules, consider using 'TestIpAddress'. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS100)

File: Trappist/src/Promact.Trappist.DomainModel/Models/Test/Test.cs [hunk 5]
```diff
@@ -33,6 +31,7 @@ public class Test : BaseModel
         public virtual ICollection<TestAttendees> TestAttendees { get; set; }
         public virtual ICollection<TestQuestion> TestQuestion { get; set; }
         public virtual ICollection<TestCategory> TestCategory { get; set; }
+        public virtual ICollection<TestIpAddress> TestIPAddress { get; set; }
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.DomainModel/Models/Test/Test.cs]: MINOR Rename property 'TestIPAddress' to match camel case naming rules, consider using 'TestIpAddress'. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS100)

File: Trappist/src/Promact.Trappist.DomainModel/Models/Test/TestIpAddress.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+﻿
+using System.ComponentModel.DataAnnotations.Schema;
+
+
+namespace Promact.Trappist.DomainModel.Models.Test
+{
+    public class TestIpAddress : BaseModel
+    {
+        public string IPAddress { get; set; }
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.DomainModel/Models/Test/TestIpAddress.cs]: MINOR Rename property 'IPAddress' to match camel case naming rules, consider using 'IpAddress'. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS100)

File: Trappist/src/Promact.Trappist.DomainModel/DbContext/TrappistDbContext.cs [hunk 7]
```diff
@@ -75,6 +75,7 @@ protected override void OnConfiguring(DbContextOptionsBuilder optionBuilder)
         public DbSet<TestAnswers> TestAnswers { get; set; }
         public DbSet<AttendeeAnswers> AttendeeAnswers { get; set; }
         public DbSet<Report> Report { get; set; }
+        public DbSet<TestIpAddress> TestIPAddress { get; set; }
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.DomainModel/DbContext/TrappistDbContext.cs]: MINOR Rename property 'TestIPAddress' to match camel case naming rules, consider using 'TestIpAddress'. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS100)

**Comment by chintans** [comment]
SonarQube analysis reported 7 issues
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: Trappist/src/Promact.Trappist.Repository/TestConduct/TestConductRepository.cs [hunk 8]
```diff
@@ -84,9 +86,14 @@ public TestConductRepository(TrappistDbContext dbContext, ITestsRepository testR
         public async Task<bool> IsTestLinkExistForTestConductionAsync(string magicString)
         {
             var testObject = await _dbContext.Test.FirstOrDefaultAsync(x => x.Link == magicString);
-            var currentDate = DateTime.UtcNow; 
-            // if Test is not paused and current date is not greater than EndDate of a test then it returns true and test link exist otherwise link does not exist
-            return testObject != null && DateTime.Compare(currentDate, testObject.EndDate) < 0 && !testObject.IsPaused ? true : false;          
+            var currentDate = DateTime.UtcNow;
+            var hostName = Dns.GetHostName();
```
Comment (rushi216) [inline Trappist/src/Promact.Trappist.Repository/TestConduct/TestConductRepository.cs]: Which ip will it return? is it current server ip or user request ip?

Comment (asif-khan17) [inline Trappist/src/Promact.Trappist.Repository/TestConduct/TestConductRepository.cs]: current server ip

Comment (rushi216) [inline Trappist/src/Promact.Trappist.Repository/TestConduct/TestConductRepository.cs]: why do you need current server ip?

**Pull request merged**